File: app/packet_builder.py
"""C013 제어 패킷 빌드.

필드명 → (code, value) 매핑은 `_items_from_fields()` 한 곳에만 존재한다.
개별 build_set_* 는 필드 dict를 구성해 이를 통과시키는 얇은 래퍼이며,
기류 연동 필드의 상호배타 규칙은 protocol.normalize_airflow()가 단일 구현이다.
"""
from __future__ import annotations

from packet_parser import crc16_xmodem, is_physical_address
from protocol import FAN_CODES, MODE_CODES, normalize_airflow

START_BYTE = 0x32
END_BYTE   = 0x34
SRC        = bytes([0x62, 0x00, 0x00])
MSG_TYPE   = bytes([0xC0, 0x13])

_seq = 0


def _next_seq() -> int:
    global _seq
    val = _seq
    _seq = (_seq + 1) & 0xFF
    return val


def _build(dst: bytes, items: list[tuple[int, bytes]]) -> bytes:
    """items: [(code_int, value_bytes), ...]"""
    # 와일드카드 주소(예: 20.ff.ff)로 보내면 모든 실내기에 동시 기록된다.
    # 유령 유닛이 어떤 경로로든 등록됐을 때의 최후 방어선.
    if not is_physical_address(dst):
        raise ValueError(f"refusing to send to non-physical address: {dst.hex()}")
    seq   = _next_seq()
    count = len(items)
    data  = b"".join(c.to_bytes(2, "big") + v for c, v in items)
    inner = SRC + dst + MSG_TYPE + bytes([seq, count]) + data
    crc   = crc16_xmodem(inner)
    size  = len(inner) + 4  # SIZE = total_packet_len - 2
    return (bytes([START_BYTE])
            + size.to_bytes(2, "big")
            + inner
            + crc.to_bytes(2, "big")
            + bytes([END_BYTE]))


# 필드명 → 인코더. 항목 순서가 곧 패킷 내 코드 순서(canonical order)다.
_FIELD_ENCODERS: list[tuple[str, callable]] = [
    ("power",           lambda v: (0x4000, bytes([0x01 if v else 0x00]))),
    ("mode",            lambda v: (0x4001, bytes([MODE_CODES.get(v, 1)]))),
    ("target_temp",     lambda v: (0x4201, int(v * 10).to_bytes(2, "big"))),
    ("fan_mode",        lambda v: (0x4006, bytes([FAN_CODES.get(v, 0)]))),
    ("vane_vertical",   lambda v: (0x4011, bytes([0x01 if v else 0x00]))),
    ("vane_horizontal", lambda v: (0x407E, bytes([0x01 if v else 0x00]))),
    ("wind_free",       lambda v: (0x4060, bytes([0x09 if v else 0x00]))),
    ("long_wind",       lambda v: (0x4007, bytes([0x10 if v else 0x0E]))),
    ("auto_clean",      lambda v: (0x4111, bytes([0x01 if v else 0x00]))),
]


def _items_from_fields(fields: dict) -> list[tuple[int, bytes]]:
    """AcStatus 필드 dict → C013 (code, value) 목록. 매핑의 단일 지점."""
    return [enc(fields[name]) for name, enc in _FIELD_ENCODERS if name in fields]


def build_fields(dst: bytes, fields: dict) -> bytes:
    """필드 dict를 C013 한 패킷으로 빌드. 빈 dict면 b''."""
    items = _items_from_fields(fields)
    return _build(dst, items) if items else b""


def build_set_power(dst: bytes, on: bool, target_temp: float | None = None) -> bytes:
    fields: dict = {"power": on}
    if on and target_temp is not None:
        fields["target_temp"] = target_temp
    return build_fields(dst, fields)


def build_set_mode(dst: bytes, mode: str) -> bytes:
    return build_fields(dst, {"mode": mode})


def build_set_target_temp(dst: bytes, temp: float, power: bool = True) -> bytes:
    return build_fields(dst, {"power": power, "target_temp": temp})


def build_set_fan_mode(dst: bytes, fan: str) -> bytes:
    return build_fields(dst, {"fan_mode": fan})


def build_set_vane(dst: bytes, vertical: bool, horizontal: bool) -> bytes:
    return build_fields(dst, normalize_airflow(
        {"vane_vertical": vertical, "vane_horizontal": horizontal}))


def build_set_wind_free(dst: bytes, on: bool) -> bytes:
    return build_fields(dst, normalize_airflow({"wind_free": on}))


def build_set_long_wind(dst: bytes, on: bool) -> bytes:
    return build_fields(dst, normalize_airflow({"long_wind": on}))


def build_set_auto_clean(dst: bytes, on: bool) -> bytes:
    return build_fields(dst, {"auto_clean": on})


def build_reconcile(dst: bytes, diffs: dict) -> bytes:
    """desired와 reported의 차이(diffs)를 C013 한 패킷으로 빌드."""
    return build_fields(dst, diffs)

File: Android/app/src/main/java/com/samsung/ac/bridge/protocol/PacketBuilder.kt
package com.samsung.ac.bridge.protocol

object PacketBuilder {
    private const val START_BYTE = 0x32
    private const val END_BYTE = 0x34
    private val SRC = byteArrayOf(0x62.toByte(), 0x00.toByte(), 0x00.toByte())
    private val MSG_TYPE = byteArrayOf(0xC0.toByte(), 0x13.toByte())

    private var seq: Int = 0

    private fun nextSeq(): Int {
        val val0 = seq
        seq = (seq + 1) and 0xFF
        return val0
    }

    private fun build(dst: ByteArray, items: List<Pair<Int, ByteArray>>): ByteArray {
        val seq = nextSeq()
        val count = items.size
        val data = items.flatMap { (code, value) ->
            code.toBytes2Big() + value.toList()
        }.toByteArray()

        val inner = SRC + dst + MSG_TYPE + byteArrayOf(seq.toByte(), count.toByte()) + data
        val crc = CRC16.xmodem(inner)
        val size = inner.size + 4

        return byteArrayOf(START_BYTE.toByte()) +
                size.toBytes2Big() +
                inner +
                crc.toBytes2Big() +
                byteArrayOf(END_BYTE.toByte())
    }

    fun buildFields(dst: ByteArray, fields: Map<String, Any?>): ByteArray {
        val items = fieldsToItems(fields)
        return if (items.isEmpty()) byteArrayOf() else build(dst, items)
    }

    fun buildSetPower(dst: ByteArray, on: Boolean, targetTemp: Float? = null): ByteArray {
        val fields = mutableMapOf<String, Any?>("power" to on)
        if (on && targetTemp != null) {
            fields["target_temp"] = targetTemp
        }
        return buildFields(dst, fields)
    }

    fun buildSetMode(dst: ByteArray, mode: String): ByteArray {
        return buildFields(dst, mapOf("mode" to mode))
    }

    fun buildSetTargetTemp(dst: ByteArray, temp: Float, power: Boolean = true): ByteArray {
        return buildFields(dst, mapOf("power" to power, "target_temp" to temp))
    }

    fun buildSetFanMode(dst: ByteArray, fan: String): ByteArray {
        return buildFields(dst, mapOf("fan_mode" to fan))
    }

    fun buildSetVane(dst: ByteArray, vertical: Boolean, horizontal: Boolean): ByteArray {
        val normalized = normalizeAirflow(mapOf("vane_vertical" to vertical, "vane_horizontal" to horizontal))
        return buildFields(dst, normalized)
    }

    fun buildSetWindFree(dst: ByteArray, on: Boolean): ByteArray {
        val normalized = normalizeAirflow(mapOf("wind_free" to on))
        return buildFields(dst, normalized)
    }

    fun buildSetLongWind(dst: ByteArray, on: Boolean): ByteArray {
        val normalized = normalizeAirflow(mapOf("long_wind" to on))
        return buildFields(dst, normalized)
    }

    fun buildSetAutoClean(dst: ByteArray, on: Boolean): ByteArray {
        return buildFields(dst, mapOf("auto_clean" to on))
    }

    fun buildApplySettings(dst: ByteArray, settings: Map<String, Any?>): ByteArray {
        val normalized = normalizeAirflow(settings)
        return buildFields(dst, normalized)
    }

    private fun normalizeAirflow(fields: Map<String, Any?>): Map<String, Any?> {
        val out = fields.toMutableMap()
        when {
            out["wind_free"] == true -> {
                out["long_wind"] = false
                out["vane_vertical"] = false
                out["vane_horizontal"] = false
            }

            out["long_wind"] == true -> {
                out["wind_free"] = false
                out["vane_vertical"] = false
                out["vane_horizontal"] = false
            }

            out["vane_vertical"] == true || out["vane_horizontal"] == true -> {
                out["wind_free"] = false
                out["long_wind"] = false
            }
        }
        return out
    }

    private fun fieldsToItems(fields: Map<String, Any?>): List<Pair<Int, ByteArray>> {
        val items = mutableListOf<Pair<Int, ByteArray>>()

        fields["power"]?.let { v ->
            val byte = if (v as Boolean) 0x01 else 0x00
            items.add(0x4000 to byteArrayOf(byte.toByte()))
        }

        fields["mode"]?.let { v ->
            val code = PacketProtocol.MODE_CODES[v as String] ?: 1
            items.add(0x4001 to byteArrayOf(code.toByte()))
        }

        fields["target_temp"]?.let { v ->
            val temp = ((v as Number).toFloat() * 10).toInt()
            items.add(0x4201 to temp.toBytes2Big())
        }

        fields["fan_mode"]?.let { v ->
            val code = PacketProtocol.FAN_CODES[v as String] ?: 0
            items.add(0x4006 to byteArrayOf(code.toByte()))
        }

        fields["vane_vertical"]?.let { v ->
            val byte = if (v as Boolean) 0x01 else 0x00
            items.add(0x4011 to byteArrayOf(byte.toByte()))
        }

        fields["vane_horizontal"]?.let { v ->
            val byte = if (v as Boolean) 0x01 else 0x00
            items.add(0x407E to byteArrayOf(byte.toByte()))
        }

        fields["wind_free"]?.let { v ->
            val byte = if (v as Boolean) 0x09 else 0x00
            items.add(0x4060 to byteArrayOf(byte.toByte()))
        }

        fields["long_wind"]?.let { v ->
            val byte = if (v as Boolean) 0x10 else 0x0E
            items.add(0x4007 to byteArrayOf(byte.toByte()))
        }

        fields["auto_clean"]?.let { v ->
            val byte = if (v as Boolean) 0x01 else 0x00
            items.add(0x4111 to byteArrayOf(byte.toByte()))
        }

        return items
    }
}

private fun Int.toBytes2Big(): ByteArray {
    return byteArrayOf(
        ((this shr 8) and 0xFF).toByte(),
        (this and 0xFF).toByte()
    )
}
